Popover: remove click-outside and rewrite the lifecycle logic (try 2)

jsnajdr opened the pull request:
Second attempt at landing the `Popover` refactoring from #37766 that was reverted in #38229.

The problem was that the deferred (`_.defer`) calls to set position after component rerender (`componentDidUpdate`) were not cancelled when the `Popover` component was hidden or unmounted immediately after that rerender. I.e., before the deferred call had a chance to run. Then the deferred callback tried to access a DOM element ref that no longer existed.

I know about two practical cases where this happens:

**Stats Chart:** when hovering over an empty chart bar, the popover is shown on `mouseenter` event, then shows right under the mouse cursor, which triggers a `mouseleave` event and the popover is hidden. Very quickly. That's a preexisting bug that triggers the "reposition after hide" condition.

Here's a screencast that shows the "appear-under-cursor" phenomenon. The hiding event is disabled:
![Screen Capture on 2019-12-06 at 09-24-28](https://user-images.githubusercontent.com/664258/70414160-8987a080-1a59-11ea-8b20-29fa11ba7272.gif)

**Opening Happychat:** Here, click on a "Chat with us" button dispatches a Redux action that opens Happychat, and the Inline Help popover also listens for change in the result of `isHappyChat` Redux selector and closes itself when Happychat appears:
```js
onButtonClick = () => {
  dispatch( openHappychat() );
}

componentWillReceiveProps( nextProps ) {
  if ( ! this.props.isHappychatOpen && nextProps.isHappychatOpen ) {
    this.closeInlineHelp();
  }
}
```
Again, rerender of the `InlineHelpPopover` component and its unmount happen very quickly after each other.

In the case of Inline Help Popover, however, I'm not 100% sure what exactly is happening. When reasoning about the code, I don't see why crash of the deferred call should prevent opening Happychat. But I can't reproduce the Happychat-not-opening bug on the updated branch.

**How I fixed it:** In the `Popover` component, I added code that carefully cancels all outstanding deferred calls when the `Popover` is hidden or unmounted.

Also, in `InlineHelpPopover`, I moved the code that closes the popover from `componentWillReceiveProps` (deprecated method) to `componentDidUpdate` (more appropriate for firing side effects). That should be another safety belt that ensures that the lifecycles execute properly.

jsnajdr committed:
Popover: remove click-outside and rewrite the lifecycle logic (try 2)

Second attempt at landing the `Popover` refactoring from #37766 that was reverted in #38229.

jsnajdr committed:
Popover: carefully cancel deferred tasks when hiding/unmounting the component

The `Popover` component uses `_.defer` to delay certain tasks, like reposition on
rerender or focus after show, to the next event loop ticks.

However, in certain edge cases (that happen often enough) the `Popover` can be hidden
or unmounted before these tasks have a chance to run. In such case, they try to work
with a DOM element that no longer exist and can crash.

This patch carefully tracks these deferred tasks and cancels them when the `Popover`
is being hidden or unmounted.

jsnajdr committed:
Inline Help: use componentDidUpdate to close help popover when opening chat

The Inline Help Popover automatically closes when Happychat opens. This patch
moves the code that does the detection and closing from the legacy `componentWillReceiveProps`
to `componentDidUpdate`, which is more appropriate for firing side effects.

Accompanies the fix for `Popover` update scheduling.

blowery commented:
When changing the site in the help popover, the popover closes when you pick a site.

![disappear](https://user-images.githubusercontent.com/14350/70444113-62d06680-1a67-11ea-8384-8757c0fce1f9.gif)

jsnajdr commented:
> When changing the site in the help popover, the popover closes when you pick a site.

There are two popovers nested in each other 🤔 And clicking inside the inner one means clicking outside the outer one. I'll check why the previous version handled this correctly.

Thanks for testing and catching this!

jsnajdr committed:
Popover: run the clickout handler in capture phase

jsnajdr commented:
@blowery That sites dropdown inside inline help that closes the popover when it shoudn't -- that's a great catch!

First, there are no nested `Popover` components as I originally thought. The `SitesDropdown` component doesn't use `Popover` and React portals, but the expanded sites list is simply an absolutely positioned `div`.

After selecting a site inside the dropdown, there are two `click` listeners that are fired:
1. There is a clickout handler on the `InlineHelpPopover`, bound to `document.body`, that checks if the `event.target` is inside the `Popover`'s DOM element and closes the popover if it isn't. `popoverEl.contains( event.target )` is used to check that.
2. There is a React click handler on the `Site` component, handling the actual business logic of selecting a site, and it also closes the `SitesDropdown`, **removing** the DOM element you just clicked on. Note also that because of React synthetic event system, the DOM handler for this click is also on `document.body`.

In the original `Popover` version, the above two handlers ran exactly in that order. First the clickout handler, then the React click handler. That's because the clickout handler was registered first. It happened in the `click-outside` module -- [see here](https://github.com/TooTallNate/click-outside/blob/master/index.js#L53-L60) how the handler is attached during static module initialization.

But with the new `Popover`, I removed the `clickout-outside` package and instead register the clickout handler during `Popover` mount. Then the two `document.body` click handlers run in a reversed order.

First, the site selection is handled and the `event.target` element gets removed from DOM (`SitesDropdown` closes)

Second, the clickout handler runs and the `popoverEl.contains( event.target )` condition is suddenly `false`. The `event.target` is no longer in the DOM tree! Click outside the `InlineHelpPopover` is detected and the popover is closed.

I fixed this bug in https://github.com/Automattic/wp-calypso/pull/38279/commits/36deb026f268845b2684ff02913ceef8bfbd3e62 by running the clickout listener in the capture phase -- before any other events (that run in the bubble phase). The right order is established again.

Please test and review 🙂

jsnajdr committed:
Popover: extract the logic that syncs isVisible prop and show state

The code that syncs the `isVisible` prop and `show` state, possibly delayed by
`showDelay` timeout, is extracted to the outer component, simplifying the inner
implementation a lot (removes approx 50 lines of code).

Also ensures that the `RootChild` is rendered (and its `div` element appended to DOM) only
after the `Popover` becomes visible. That ensures that when the last of several nested
`Popover`s is shown, it's also the last DOM element in the DOM tree, and therefore has
higher natural z-index and is displayed over the earlier one.

jsnajdr commented:
Today I fixed one more subtle bug that I discovered when testing @sgomes' Publicize changes in #37078. There are two nested popovers there:

<img width="246" alt="Screenshot 2019-12-17 at 13 19 29" src="https://user-images.githubusercontent.com/664258/70995762-9a778800-20d1-11ea-8e92-8bed64fc2bf8.png">

And the render code looks somewhat like this:
```jsx
const [ showOuterPopover, setOuterPopover ] = useState( false );
const [ showInnerPopover, setInnerPopover ] = useState( false );
const anchorRef = useRef();

return (
  <>
    { showOuterPopover && <Popover>
        <OuterPopoverContent
          anchorRef={ anchorRef }
          toggleInnerPopover={ setInnerPopover }
        />
      </Popover> }
    <Popover isVisible={ showInnerPopover } context={ anchorRef.current } >
      <InnerPopoverContent />
    </Popover>
  </>
);
```

The `OuterPopoverContent` sets the `anchorRef` to a DOM element inside itself to anchor the inner popover to, and also controls whether the inner popover is displayed.

However, when mounting this UI, the inner popover is mounted first, because its render is not conditional, but rather uses the `isVisible` prop. And because its portal and the `div` container is rendered even for `isVisible` false, the `div` for the inner popover is in the `document.body` before the `div` for the outer popover. And therefore has a lower natural z-index and is shown behind it :-1:

I refactored the `Popover` in such a way that the portal (i.e., the `RootChild` component) is rendered only after `isVisible` becomes `true`. That removes the difference between conditional and unconditional render and the popover that shows last is the last one in DOM, and also on top of the z-index order.

blowery reviewed (approved):
LGTM! Stats, the contact us popover, and the post scheduling popover all work as expected.
